### DISCUSSION TITLE
[How-to] Parsing a specific regex to a custom tag?

### DISCUSSION THREAD

**Discussion by rafaell-lycan** [ueberdosis/tiptap#4171] [General Questions]
Hi y'all 👋 

I'm doing research on our next WYSIWYG Editor and _Tiptap_ has checked almost all the boxes.

I've used it on a personal project, but at my company, we have Quill with a bunch of in-house plugins to support our needs, which is getting harder to support and maintain.

Essentially we have a parser on our application that adds and reads a [mustache template](https://en.wikipedia.org/wiki/Mustache_(template_system)) and shows it as a react/html component.

We do that using a Regex: `/{{([a-zA-Z0-9]+):([a-zA-Z0-9-_]+)}}/g`

e.g.:
```bash
# Syntax
{{var:name}}

# A link that uses variables
https://tiptap.dev/?tag={{var:finalending}}&score={{var:score}}

```

Which renders something like this:

<img width="394" alt="Screenshot 2023-06-30 at 11 35 27" src="https://github.com/ueberdosis/tiptap/assets/1516450/20c510e2-095b-4d56-bc72-a51c8cb10997">

After trying to use the [Suggestion](https://tiptap.dev/api/utilities/suggestion) pkg, I wasn't able to use it as seamlessly as the current implementation we have as it uses HTML tags underneath.

Does anyone have an idea of how to manage a similar scenario?

Thanks

**Comment by philipisik** [comment]
@rafaell-lycan Is there a function hidden behind the rendered labels (finalending and score), e.g. should a context menu open when clicking or is it purely for display in the URL?

**Reply by rafaell-lycan** [reply]
Hi @philipisik 

Our use case is essentially something very close to the Suggestion plugin/Mention extension, meaning how we present data to the user when they type `@`, and after having an option rendered you can also remove it (also similar to the Suggestion/Mention behavior):

<img width="603" alt="Screenshot 2023-06-30 at 17 29 43" src="https://github.com/ueberdosis/tiptap/assets/1516450/bf97f6ed-2fb1-44b6-b785-e4ac1f556981">


I hope this can clarify a bit the flow of this particular usage.

**Comment by svenadlung** [comment]
That should be possible with Tiptap - but might require an individual solution. 
It's not a one-size-fits-all approach and depends on a number of factors and your exact requirements. Feel free to reply to @philipisik if you need further assistance. Happy to help.

**Comment by rafaell-lycan** [comment]
Hey @philipisik @svenadlung 👋 

This is a solution we are trying to reimplement at @Typeform and this scenario is one of the last missing features to migrate from Quill to Tiptap.

We store the content as markdown and parse it later without transforming the mustache syntax having something like these:

Giving a Markdown (API format):
```md
**Sweet!**
Your score is: {{var:score}}

You can check your answers [here](http://example.com/{{var:custom_id}}?utm_source={{var:utm_source}}).
```

HTML result (Editor input):
```html
<p><strong>Sweet!</strong><br>Your score is: {{var:score}}<br><br><a target="_blank" rel="noopener noreferrer nofollow" href="http://example.com/{{var:custom_id}}?utm_source={{var:utm_source}}">here</a>.</p>
```

We have some code that does the conversion when matching that Regex and converts to code that has one appearance on the editor but changes when converting the output. And after checking the Suggestion/Mention, I understood that it won't work with links as it is a `node` type, and I presume could only work if that was a `mark` instead (as it needs to interpolate with links for example).

By simply using the current plugin it generates the following HTML when converting to this, and the `@` trigger won't work inside of a HTML tag.

```html
<p><span data-type="mention" class="mention">@score</span></p>
```

This generates a `node`, it won't be possible to interpolate inside of a link AFAIK.

@svenadlung do you see/have any example where it would work for this scenario?

Thanks a lot.

cc @sh-waqar
